com.unity.shadergraph/Editor/ShaderFoundry/Linker/Properties/PropertyDeclarations.cs

```diff
@@ -0,0 +1,382 @@
+            var accessorFnName = "UnityBuildTexture2DStruct";
+            if (!hasScaleOffset)
+                accessorFnName = "UnityBuildTexture2DStructNoScale";
+            var buildingData = new TextureBuildingData
+            {
+                TextureUniformType = context.Container._Texture2D,
+                MaterialPropertyType = "2D",
+                AssignmentFunctionName = accessorFnName,
+                HasTexelSize = true,
+                HasScaleOffset = hasScaleOffset,
+            };
+            return HandleUnityTextureShared(context, buildingData, result);
+        }
+
+        static bool HandleUnityTexture2DArray(FieldPropertyContext context, FieldPropertyData result)
+        {
+            var buildingData = new TextureBuildingData
+            {
+                TextureUniformType = context.Container._Texture2DArray,
+                MaterialPropertyType = "2DArray",
+                AssignmentFunctionName = "UnityBuildTexture2DArrayStruct",
+            };
+            return HandleUnityTextureShared(context, buildingData, result);
+        }
+
+        static bool HandleUnityTextureCube(FieldPropertyContext context, FieldPropertyData result)
+        {
+            var buildingData = new TextureBuildingData
+            {
+                TextureUniformType = context.Container._TextureCube,
+                MaterialPropertyType = "CUBE",
+                AssignmentFunctionName = "UnityBuildTextureCubeStruct",
+            };
+            return HandleUnityTextureShared(context, buildingData, result);
+        }
+
+        static bool HandleUnityTexture3D(FieldPropertyContext context, FieldPropertyData result)
+        {
+            var buildingData = new TextureBuildingData
+            {
+                TextureUniformType = context.Container._Texture3D,
+                MaterialPropertyType = "3D",
+                AssignmentFunctionName = "UnityBuildTexture3DStruct",
+            };
+            return HandleUnityTextureShared(context, buildingData, result);
+        }
+
+        internal class TextureBuildingData
+        {
+            internal ShaderType TextureUniformType;
+            internal string MaterialPropertyType;
+            internal string MaterialPropertyDefaultValue = "\"\" {}";
+            internal string AssignmentFunctionName;
+            internal bool HasScaleOffset = false;
+            internal bool HasTexelSize = false;
+        }
+
+        static bool HandleUnityTextureShared(FieldPropertyContext context, TextureBuildingData buildingData, FieldPropertyData result)
+        {
+            var container = context.Container;
+            var uniformName = context.UniformName;
+            result.UniformReadingData = new UniformReadingData { Rhs = $"{buildingData.AssignmentFunctionName}({uniformName})" };
+            result.UniformDeclarations.Add(new UniformDeclarationData { dataSource = context.DataSource, Name = $"{uniformName}", Type = buildingData.TextureUniformType });
```

Comment:
Do we need to wrap these in macros here or is it done elsewhere?

Reply:
Yes, the building of the "Unity" resource objects goes through macros so it doesn't change across all platforms.